### ISSUE TITLE
hello , I want to know the MS COCO online test .

### ISSUE THREAD

**Issue by PanYuQi66666666** [jchenghu/ExpansionNet_v2#14]
When I was doing the MS COCO online test, I encountered an error :   annsImgIds = [ann['image_id'] for ann in anns] KeyError: 'image_id' ,Could you please give me a tutorial for doing mscoco online test? I have been trying to do it for a long time but failed. When doing offline test, should I store image_id and caption fields in json file and then upload them for testing? Does image_id refer to the name of the picture?

**Comment by jchenghu** [comment]
Hi, the code should already be available in `test.py` for offline testing (I'm assuming you mean the offline test set according to Karpathy's split). 

Regarding the online testing, you should create a JSON file structured in this way:

```
[{"caption": "a man standing in a field holding a large kite", "image_id": 503167}, {"caption": "a group of motorcycles parked in a building", "image_id": 451615}, {"caption": "a woman cutting her hair with a woman s hair", "image_id": 531214}, {"caption": "two people and a dog on a boat in the water", "image_id": 360107}, {"caption": "a brown and white dog sitting under a park bench", "image_id": 56757}]
```

This is only a small example, but the version required by the online evaluation server contains about 40k entries and the images and image_ids can be downloaded here:

![image](https://github.com/user-attachments/assets/397f6f2a-035d-4789-a2a3-b98014adf208)

Once the download is complete, you can do something like this, create a custom version of test.py and here:

```
def test(
	...
	):
    print("GPU: " + str(rank) + "] Process " + str(rank) + " working...")

    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = ddp_sync_port
    dist.init_process_group("nccl", rank=rank, world_size=world_size)

	...

    model.to(rank)
    ddp_model = DDP(model, device_ids=[rank])

    data_loader = CocoDataLoader(coco_dataset=coco_dataset,
                                 batch_size=1,
                                 num_procs=world_size,
                                 array_of_init_seeds=array_of_init_seeds,
                                 dataloader_mode='image_wise',
                                 resize_image_size=img_size if is_end_to_end else None,
                                 rank=rank,
                                 verbose=False)

    if not is_ensemble:
        print("Not ensemble")
        map_location = {'cuda:%d' % 0: 'cuda:%d' % rank}
        checkpoint = torch.load(save_model_path, map_location=map_location)
        model.load_state_dict(checkpoint['model_state_dict'], strict=is_end_to_end)
    else:
        print("Ensembling Evaluation")
        list_checkpoints = os.listdir(save_model_path)
        checkpoints_list = [save_model_path + elem for elem in list_checkpoints if elem.endswith('.pth')]
        print("Detected checkpoints: " + str(checkpoints_list))

        if len(checkpoints_list) == 0:
            print("No checkpoints found")
            dist.destroy_process_group()
            exit(-1)
        ddp_model = get_ensemble_model(model, checkpoints_list, rank=rank)


    < < < < ADD/REPLACE CODE HERE > > > >

    print("[GPU: " + str(rank) + " ] Closing...")
    dist.destroy_process_group()

```

in  < < < < ADD/REPLACE CODE HERE > > > > put the following:

```
    ids_list = < < < the complete LIST of img_ids > > > 
    # this should be contained in the online test2014 annotations
    
    ddp_model.eval()

    chunk_size = parallel_batches
    num_chunk = math.ceil(len(ids_list) / chunk_size)

    strings_to_be_written_on_file = []

    # ----------------------------------------------------------
    from PIL import Image as PIL_Image
    import torchvision
    import json

    def generate_tensor_from_image(file_path):
        pil_image = PIL_Image.open(file_path)
        if pil_image.mode != 'RGB':
            pil_image = PIL_Image.new("RGB", pil_image.size)
        preprocess_pil_image = test_preprocess_1(pil_image)
        tens_image = torchvision.transforms.ToTensor()(preprocess_pil_image)  # [channel, height, width]
        tens_image = test_preprocess_2(tens_image)
        return tens_image

    test_preprocess_layers_1 = [torchvision.transforms.Resize((384, 384))]
    test_preprocess_layers_2 = [torchvision.transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])]
    test_preprocess_1 = torchvision.transforms.Compose(test_preprocess_layers_1)
    test_preprocess_2 = torchvision.transforms.Compose(test_preprocess_layers_2)

    def convert_id_to_img_path(id, image_path):
        str_id = str(id)
        len_id = len(str_id)
        num_pads = 12 - len_id
        zero_pads = ''.join(['0'] * num_pads)
        final_name = image_path + 'COCO_test2014_' + zero_pads + str_id + '.jpg'
        return final_name
    # ---------------------------------------------

    
    json_pred_list = []
    for it in range(num_chunk):
        from_idx = it * chunk_size
        to_idx = min((it + 1) * chunk_size, len(ids_list))

        list_predictions = []

        list_tensor = [generate_tensor_from_image(
                                convert_id_to_img_path(id, '/mnt/ssd_data/MS_COCO_2014/test2014/')
                       ) for id in ids_list[from_idx:to_idx]]
        sub_batch_x = torch.stack(list_tensor, dim=0).to(rank)
        sub_batch_x_num_pads = [0] * sub_batch_x.size(0)
        #sub_batch_x = torch.nn.utils.rnn.pad_sequence(validate_x, batch_first=True).to(rank)
        #sub_batch_x_num_pads = compute_num_pads(validate_x)

        beam_search_kwargs = {'beam_size': beam_size,
                              'beam_max_seq_len': 30, # max_seq_len,
                              'sample_or_max': 'max',
                              'how_many_outputs': 1,
                              'sos_idx': sos_idx,
                              'eos_idx': eos_idx}

        output_words, _ = ddp_model(enc_x=sub_batch_x,
                                    enc_x_num_pads=sub_batch_x_num_pads,
                                    mode='beam_search', **beam_search_kwargs)

        del sub_batch_x

        # take the first element of beam_size sequences
        output_words = [output_words[i][0] for i in range(len(output_words))]

        pred_sentence = language_utils.convert_allsentences_idx2word(output_words, y_idx2word_list)
        for sentence in pred_sentence:
            list_predictions.append(' '.join(sentence[1:-1]))  # remove EOS and SOS

        for id, caption in zip(ids_list[from_idx:to_idx], list_predictions):
            json_pred_list.append([{'image_id': id, 'caption': caption})
      
    with open('./my_coco2014_predictions.json', 'w') as f:
        json.dump(json_pred_list, f)    


```

this code is not parallelized, but since you need to do that once It shouldn't be too much of a problem.
It should work, but I suggest you test it on a few img_ids before running the complete test (which should take a few hours) first.

Now, in the code above, I called the results 'my_coco2014_predictions.json', simply rename it into `captions_test2014_<yourname>_results.json`.

Now prepare a zip file containing `captions_test2014_<yourname>_results.json` and `captions_val2014_<yourname>_results.json` name it `results.zip`
this is the file to be submitted to the codalab server.

The server asks also for the validation results, the procedure is the same as the above but with the val2014 images... However, you might want simply to create a copy of `captions_test2014_<yourname>_results.json` since it does not impact the test2014 online evaluation and I think it is only an optional tool for you to ensure your evaluation implementation coincides with theirs.



That should be it, I hope it helps

**Comment by PanYuQi66666666** [comment]
I used the test set and validation set divided by offline testing directly, saved them as two json files, and sent them to execute, but it still reported an error. Is there something wrong with my operation?
![屏幕截图 2024-08-29 145425](https://github.com/user-attachments/assets/20ff9a26-c117-4b02-968b-56aa898dcc17)

**Comment by PanYuQi66666666** [comment]
![image](https://github.com/user-attachments/assets/ae85d81f-247c-48fd-ba56-a9542852c99b)
Please tell me how to fill in this field.

**Comment by PanYuQi66666666** [comment]
Could you please provide a complete code for reference?

**Comment by jchenghu** [comment]
Try these

[test2014_ids.txt](https://github.com/user-attachments/files/16801554/test2014_ids.txt)
[test_coco2014.txt](https://github.com/user-attachments/files/16801555/test_coco2014.txt)

rename them from ".txt" to ".py" files (github blocks "py" files)
then try this command

```
python test_coco2014.py --N_enc 3 --N_dec 3 --model_dim 512 \
    --num_gpus 1 --eval_beam_sizes [5] --is_end_to_end True \
    --eval_parallel_batch_size 4 \
    --test2014_images_path < < YOUR_FOLDER OF TEST2014 images installation> >/test2014/  \
    --images_path ./github_ignore_material/raw_data/<your_TRAINING_coco_img_folder> \
    --save_model_path ./github_ignore_material/saves/phase6_checkpoint
```

to perform a quick test, remove as many elements as you want in test2014_ids.py,
let me if it works, I cannot try it myself at the moment since I don't have the coco dataset on my current machine

Edit: for easier debugging, you can put a print in line 215 of file test_coco2014.py:
```
for id, caption in zip(ids_list[from_idx:to_idx], list_predictions):
            json_pred_list.append([{'image_id': id, 'caption': caption})
            print(str(id) + ': ' + caption)  # <-- this line
```

**Comment by PanYuQi66666666** [comment]
Thanks! I will try it .

**Comment by jchenghu** [comment]
Hi, I assume you solved this because of the other thread, feel free to re-open it if that's not the case
